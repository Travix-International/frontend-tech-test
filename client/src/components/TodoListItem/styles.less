.item-container {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  margin-top: 0.625rem;

  &-info-section h2 {
    margin: 0;
  }

  .delete-btn {
    font-size: 2rem;
    color: red;
    cursor: pointer;
    z-index: 999;
  }
}
