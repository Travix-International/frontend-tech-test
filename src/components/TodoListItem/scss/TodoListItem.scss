$breakpoint-tablet: 768px;

.taskItemContainer{
    display: flex;
    flex-direction: row;
    border-bottom: 1px solid #eaeaea;
    padding: 12px;
    align-items: center;
    max-width: 850px;

    .checkButton{
        border: 1px solid #e4e4e4;
        background-color: #fff;
        border-radius: 10px;
        width: 20px;
        height: 20px;
        box-sizing: border-box;
        padding: 0px;
        display: flex;
        justify-content: center;
        cursor: pointer;
        color: #eaeaea;
        margin-right: 8px;
        
        .checkIcon{
            display: none;
        }

        .undoIcon{
            display: none;
        }
    }
    .checkButton:hover{
        background-color: #31c37b;
        color: #fff;
        border-color: #31c37b;
        .checkIcon{
            display: block;
        }
    }

    .taskTextContainer{
        color: #686868;
        font-size: 16px;
        white-space: nowrap;
        overflow: hidden;
        text-overflow: ellipsis;
        max-width: 75%;

        a{
            text-decoration: none !important;
            color: #686868;
            transition: all ease;
        }
    }
    .taskTextContainer:hover{
        a{
            color: #000;
        }
    }

    .removeButton{
        margin-left: 6px;
        font-size: 17px;
        color: #e1e1e1;
        cursor: pointer;
    }

    .removeButton:hover{
        color: #d72727;
    }
}

.taskCompleted{
    .taskTextContainer{
        a{
            color: #d1d1d1;
            font-size: 16px;
            text-decoration: line-through !important;
        }
    }
    .taskTextContainer:hover{
        a{
            color: #d1d1d1;
        }
    }

    .checkButton{
        border: 1px solid #fff;
        background-color: #fff;
        border-radius: 10px;
        width: 20px;
        height: 20px;
        box-sizing: border-box;
        padding: 0px;
        display: flex;
        justify-content: center;
        cursor: pointer;
        color: #31c37b;
        margin-right: 8px;
        
        .checkIcon{
            display: block;
        }
    }

    .checkButton:hover{
        .checkIcon{
            display: none;
        }

        .undoIcon{
            display: block;
        }
    }
}


@media only screen and (min-width: $breakpoint-tablet){
    .taskItemContainer{
        .taskTextContainer{
            font-size: 16px;
            max-width: 90%;
        }
    }
}