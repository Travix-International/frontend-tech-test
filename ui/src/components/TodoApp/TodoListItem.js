/**
 * Created by maykinayki on 9/8/17.
 */

import React  from 'react';

export default ({item, onTodoDelete, onTodoUpdate, onTodoModeChange}) => {

    const content = item.mode === "update" ? (
        <form className="todo-form" onSubmit={(e) => {
            e.preventDefault();
            onTodoUpdate(item.id, new FormData(e.currentTarget));
        }}>
            <div className="todo-input">
                <input name="title" className="input" placeholder="Title" defaultValue={item.title} />
                <textarea name="description" className="input" autoFocus="autoFocus" placeholder="Leave a note" defaultValue={item.description} />
            </div>
            <button className="todo-btn" type="submit">Save</button>
            <button className="todo-btn cancel" type="button" onClick={()=> onTodoModeChange(item.id, "default")}>Cancel</button>
        </form>
    ) : (
        <label htmlFor="checkbox">
            <input className="todo-item--checkbox" type="checkbox" />
            <div className="todo-item">
                <div className="todo-item-inner">
                    <div className="title">{item.title}</div>
                    <div className="description">{item.description}</div>
                </div>
                <button className="todo-item--delete" onClick={()=> onTodoDelete(item.id)}>Ã—</button>
                <button className="todo-item--edit" onClick={()=> onTodoModeChange(item.id, "update")}>
                    <svg fill="#FFFFFF" height="14" viewBox="0 0 24 24" width="14" xmlns="http://www.w3.org/2000/svg">
                        <path d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.39-.39-1.02-.39-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"/>
                        <path d="M0 0h24v24H0z" fill="none"/>
                    </svg>
                </button>
            </div>
        </label>
    );

    return (
        <div className="todo-list-item">
            {content}
        </div>
    )
}