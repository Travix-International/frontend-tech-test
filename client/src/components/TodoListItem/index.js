import React from 'react'
import PropTypes from 'prop-types'

import './styles.less'

class TodoListItem extends React.PureComponent {
  deleteTodoRequest = e => {
    const { id } = this.props.todo

    this.props.deleteTodoRequest({ id })
  }

  toggleEditMode = () => {
    const { id } = this.props.todo

    this.props.toggleEditMode({ editedTodoId: id })
  }

  render() {
    const { todo } = this.props

    return (
      <div className="item-container" onDoubleClick={this.toggleEditMode}>
        <div className="item-container-info-section">
          <h2>{todo.title}</h2>
          <div>{todo.description}</div>
        </div>
        <span className="delete-btn" onClick={this.deleteTodoRequest}>
          &#215;
        </span>
      </div>
    )
  }
}

TodoListItem.propTypes = {
  deleteTodoRequest: PropTypes.func.isRequired,
  toggleEditMode: PropTypes.func.isRequired,
  todo: PropTypes.shape({
    id: PropTypes.string,
    title: PropTypes.string,
    description: PropTypes.string
  })
}

export default TodoListItem
