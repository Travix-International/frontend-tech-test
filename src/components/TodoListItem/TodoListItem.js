import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import FaCheck from 'react-icons/lib/fa/check';
import FaMinus from 'react-icons/lib/fa/minus-circle';
import FaUndo from 'react-icons/lib/fa/rotate-left';
import './scss/TodoListItem.scss';


class TodoListItem extends Component{

	constructor(props){
		super(props);
	}

	render(){
		let status = this.props.itemInfo.status;
		if(status == undefined){
			status = true;
		}
		let taskClasses = status ? "taskItemContainer" : "taskItemContainer taskCompleted"
		return (
		<div className={taskClasses}>
			<button className="checkButton" id="toggleButton" onClick={() => this.props.change(this.props.itemInfo.id)}><FaCheck className="checkIcon"/><FaUndo className="undoIcon"/></button>
			<div className="taskTextContainer"><Link to={"/task/" + this.props.itemInfo.id} >{this.props.itemInfo.title}</Link></div>
			<FaMinus className="removeButton" id="removeButton" onClick={() => this.props.remove(this.props.itemInfo.id)} />
		</div>
		);
	}
}

export default TodoListItem;