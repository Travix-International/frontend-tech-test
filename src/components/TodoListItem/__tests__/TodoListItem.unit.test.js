import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import TodoListItem from '../TodoListItem';

test('Given proper props the list item renders properly.', () => {
	const element = {id:0, task:"Sample Task Text Goes Here.", description:"Some Description Goes Here", status: undefined};
	const toggleStatus = () => {};
	const removeFromList = () => {};
	const wrapper = shallow(<TodoListItem key={element.id} itemInfo={element} change={toggleStatus} remove={removeFromList}/>);
	expect(toJson(wrapper)).toMatchSnapshot();
});

test('ToggleStatus method provided as a prop is called on checking a task as done.', () => {
	const element = {id:0, task:"Sample Task Text Goes Here.", description:"Some Description Goes Here", status: true};
	const toggleStatus = jest.fn();
	const removeFromList = () => {};
	const wrapper = shallow(<TodoListItem key={element.id} itemInfo={element} change={toggleStatus} remove={removeFromList}/>);
	wrapper.find('#toggleButton').simulate('click');
	expect(toggleStatus.mock.calls.length).toBe(1);
});

test('removeFromList method provided as a prop is called on delete task click.', () => {
	const element = {id:0, task:"Sample Task Text Goes Here.", description:"Some Description Goes Here", status: true};
	const toggleStatus = () => {};
	const removeFromList = jest.fn(); 
	const wrapper = shallow(<TodoListItem key={element.id} itemInfo={element} change={toggleStatus} remove={removeFromList}/>);
	wrapper.find('#removeButton').simulate('click');
	expect(removeFromList.mock.calls.length).toBe(1);
});