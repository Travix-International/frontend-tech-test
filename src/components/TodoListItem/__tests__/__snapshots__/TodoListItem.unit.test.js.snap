// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given proper props the list item renders properly. 1`] = `
<div
  className="taskItemContainer"
>
  <button
    className="checkButton"
    id="toggleButton"
    onClick={[Function]}
  >
    <FaCheck
      className="checkIcon"
    />
    <FaRotateLeft
      className="undoIcon"
    />
  </button>
  <div
    className="taskTextContainer"
  >
    <Link
      replace={false}
      to="/task/0"
    />
  </div>
  <FaMinusCircle
    className="removeButton"
    id="removeButton"
    onClick={[Function]}
  />
</div>
`;
